Removing viper lib from viperutil/config_util.go

nbasker opened the pull request:
#### Type of change

- Improvement (improvement to code and performance by avoiding viper library)

#### Description

There is overlap between in the implementation of
`getKeysRecursively` and the viper.Get() implementation.
The viper does yaml.Unmarshal and stores in viper.config,
which is provided to getKeysRecursively and further sub-keys
are available readily and there is no need to fetch
through viper.Get(). Towards the overall objective of
moving away from viper, its usage here can be removed.

Updated the PR with changes to remove "Viper" library from following three modules. These are the modules that were using Viper along with EnhancedExactUnmarshal()/getKeysRecursively() that had overlapping functionality.
1) orderer/common/localconfig
2) internal/configtxgen/genesisconfig
3) common/viperutil

In the common/viperutil/config_util.go introduced a ConfigParser that tracks the config-files, paths, environment variables and unmarshalled-config. It may be better to create a new directory file called common/configparser/configparser.go. Based on review comments of this, shall take up this change.

#### Additional details

Updated unit test of the three modules

go test -race -count 1 -v -run Test github.com/hyperledger/fabric/orderer/common/localconfig | tail -2
PASS
ok      github.com/hyperledger/fabric/orderer/common/localconfig        5.437s
go test -race -count 1 -v -run Test github.com/hyperledger/fabric/internal/configtxgen/genesisconfig | tail -2
PASS
ok      github.com/hyperledger/fabric/internal/configtxgen/genesisconfig        5.884s
go test -race -count 1 -v -run Test github.com/hyperledger/fabric/common/viperutil | tail -2
PASS
ok      github.com/hyperledger/fabric/common/viperutil  5.507s
go test -race -count 1 -v -run Test github.com/hyperledger/fabric/orderer/consensus/etcdraft | tail -2
PASS
ok      github.com/hyperledger/fabric/orderer/consensus/etcdraft        59.218s
go test -race -count 1 -v -run Test github.com/hyperledger/fabric/orderer/consensus/etcdraft | grep SUCCESS!
SUCCESS! -- 105 Passed | 0 Failed | 0 Pending | 0 Skipped

#### Related issues

https://jira.hyperledger.org/browse/FAB-17508

srderson commented:
README should be an md file

block-listener.go needs gofmt, go vet, golint run. Public functions require comments and they are missing in this file.

tuand27613 commented:
golint/gofmt'ed block-listener.go

README now README.md

srderson commented:
Thanks @tuand27613! LGTM

srderson commented:
Also thanks @muralisrini for the readme and sample app! These are great for people learning to use the event mechanism.

nbasker committed:
Removing viper lib from viperutil/config_util.go

There is overlap between in the implementation of
`getKeysRecursively` and the viper.Get() implementation.
The viper does yaml.Unmarshal and stores in viper.config,
which is provided to getKeysRecursively and further sub-keys
are available readily and there is no need to fetch
through viper.Get(). Towards the overall objective of
moving away from viper, its usage here can be removed.

https://jira.hyperledger.org/browse/FAB-17508
Improve the efficiency of viperutil.EnhancedExactUnmarshal

Updating with changes to remove references of viper from
a) internal/configtxgen/genesisconfig
b) orderer/common/localconfig
c) common/viperutil

It is better to change common/viperutil to common/configparse, shall
consider this with review comments on this change.

No new tests are added, the following unit tests done
go test -race -count 1 -v -run Test
github.com/hyperledger/fabric/orderer/common/localconfig | tail -2
PASS
ok      github.com/hyperledger/fabric/orderer/common/localconfig
5.437s
go test -race -count 1 -v -run Test
github.com/hyperledger/fabric/internal/configtxgen/genesisconfig | tail
-2
PASS
ok      github.com/hyperledger/fabric/internal/configtxgen/genesisconfig
5.884s
go test -race -count 1 -v -run Test
github.com/hyperledger/fabric/common/viperutil | tail -2
PASS
ok      github.com/hyperledger/fabric/common/viperutil  5.507s
go test -race -count 1 -v -run Test
github.com/hyperledger/fabric/orderer/consensus/etcdraft | tail -2
PASS
ok      github.com/hyperledger/fabric/orderer/consensus/etcdraft
59.218s
go test -race -count 1 -v -run Test
github.com/hyperledger/fabric/orderer/consensus/etcdraft | grep SUCCESS!
SUCCESS! -- 105 Passed | 0 Failed | 0 Pending | 0 Skipped

Signed-off-by: NicholasB <nbasker@gmail.com>

nbasker commented:
Manually ran the same test case a few times and it passed. Shall resubmit with further cleanup after getting review feedback.

go test -race -count 1 -v -run Test github.com/hyperledger/fabric/orderer/consensus/etcdraft 
<--snip-->                                                                                                                                                      
••2020-06-12 09:10:02.116 IST [test] HandleChain -> INFO 02b EvictionSuspicion not set, defaulting to 10m0s
•                                                                                
Ran 105 of 105 Specs in 44.319 seconds                                           
SUCCESS! -- 105 Passed | 0 Failed | 0 Pending | 0 Skipped                
--- PASS: TestEtcdraft (44.32s)

denyeart commented:
Tagging @sykesm for review since the PR is related to a Jira he opened.

sykesm reviewed (commented):
Thanks for the PR. Sorry for the delay in reviewing. I've been focusing on other things.

Looks like the approach you've taken is to extract a bunch of code from viper and inlined it into the `viperutil` package. While normally I'd push back on that because we wouldn't benefit from the enhancements and fixes from that community, but, given we're stuck on an ancient version, it's probably good enough.

My goal is to do a detailed review over the next day or two. Based on my quick glance, I think there are a few things that were pulled over that we simply don't need and that should be removed (e.g. `SetConfigType`, `AutomaticEnv`), there are some idiomatic Go changes that I'd like to see, and there are tests that need to be backfilled.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+// default config file path
+const OfficialPath = "/etc/hyperledger/fabric"
+
+var SupportedExts []string = []string{"yaml", "yml"}
```
The slice can be inlined in the range that's happening below. No need to have the exported field for it.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
 
-func getKeysRecursively(base string, getKey viperGetter, nodeKeys map[string]interface{}, oType reflect.Type) map[string]interface{} {
+// default config file path
+const OfficialPath = "/etc/hyperledger/fabric"
```
This was copied from fabric/core/config and I'm not really sure why.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+}
+
+// Creates a ConfigParser instance
+func New() *ConfigParser {
+	c := new(ConfigParser)
+	c.config = make(map[string]interface{})
+	return c
+}
```
More idiomatically:
```
return &ConfigParser{
    config: map[string]interface{}{},
}
```

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+	// env variable prefix
+	envPrefix           string
+	automaticEnvApplied bool
+	envKeyReplacer      *strings.Replacer
```
We don't need to preserve the generalizations of viper. We have one pattern and can simply inline a call to `strings.ReplaceAll` where we need to map `.` to `_`.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+
+	// env variable prefix
+	envPrefix           string
+	automaticEnvApplied bool
```
We don't need `automaticEnvApplied` or the associated `AutomaticEnv` method since we don't need the generalizations of viper.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+// Keeps a list of path to search the relevant
+// config file. Multiple paths can be provided.
+func (c *ConfigParser) AddConfigPath(in string) {
+	if in != "" {
```
No need to wrap this with a zero value check. And, if we did need one, idiomatically we would return immediately on the condition and leave the real logic as the mainline path.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+
+// The config file name. The extension is not included
+func (c *ConfigParser) SetConfigName(in string) {
+	if in != "" {
```
Sane here. No need to wrap with a zero value check.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+
+// The extension of the config file
+// Currently only "yaml" is supported.
+func (c *ConfigParser) SetConfigType(in string) {
```
No need for this method at all. We only support yaml.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+// Method to initialize default paths.
+// The defaults paths are CWD and OfficialPath
+// The env variable FABRIC_CFG_PATH can be used to override default
+func (c *ConfigParser) InitConfigPath(configName string) error {
```
I believe this should be a function instead of a method that simply returns the paths to search. There's no need to do any validation here since searching does the same thing.

Probably something as simple as

```go
func ConfigPaths() []string {
    var paths []string
    if p := os.Getenv("FABRIC_CFG_PATH"); p != "" {
        paths = append(paths, p)
    }
    return append(paths, ".", "/etc/hyperledger/fabric")
}
```

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+
+// Keeps a list of path to search the relevant
+// config file. Multiple paths can be provided.
+func (c *ConfigParser) AddConfigPath(in string) {
```
The method can be made plural and the arg list variadic. That would allow us to add multiple paths in one go.

If combined with the suggestions around removing `InitConfigPath`, we can set all paths in one go.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+// Explicitly setting the full config file name.
+// In this case, configPaths search shall not be done
+func (c *ConfigParser) SetConfigFile(in string) {
+	if in != "" {
```
No need for the zero value check

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+// If "peer" is set here, all environment variable
+// searches shall be prefixed with "PEER_"
+func (c *ConfigParser) SetEnvPrefix(in string) {
+	if in != "" {
```
no need for the zero value check

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+// Search for the existence of filename for all supported extensions
+func (c *ConfigParser) searchInPath(in string) (filename string) {
+	logger.Debugf("Searching for config in ", in)
+	for _, ext := range SupportedExts {
```
SupportedExts can be an inline slice of `yaml` and `yml`.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+		return err
+	}
+
+	//c.config = make(map[string]interface{})
```
remove commented out code, please.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+
+	cf := c.getConfigFile()
+	logger.Debugf("Attempting to read from config file : %s", cf)
+	file, err := ioutil.ReadFile(cf)
```
Please pass the file as a reader to `ReadConfig` instead of reading it fully into a slice and then creating a reader from the buffer. Not only does it make simpler, it makes it more efficient and removes the duplicated logic.

sykesm replied:
This doesn't quite address the comment. The code is still fully reading the file instead of opening it and passing the file as the reader to ReadConfig.

```go
...
    file, err := os.Open(cf)
    if err != nil {
        return err
    }
    defer file.Close()

    return c.ReadConfig(file)
```

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+	//c.config = make(map[string]interface{})
+	buf := new(bytes.Buffer)
+	buf.ReadFrom(in)
+	return yaml.Unmarshal(buf.Bytes(), c.config)
```
Please use the decoder instead of reading the entire contents of the reader into a buffer and then passing it to `Unmarshal`. By creating a decoder instance, it can stream the data.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+	return yaml.Unmarshal(buf.Bytes(), c.config)
+}
+
+// Specifies the replacement string.
+// Used in searching env. variables.
+func (c *ConfigParser) SetEnvKeyReplacer(r *strings.Replacer) {
+	c.envKeyReplacer = r
+}
+
+// Set the environment variable search boolean
+func (c *ConfigParser) AutomaticEnv() {
+	c.automaticEnvApplied = true
+}
+
```
No need for these. We don't need the generalizations of viper.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+
+// Get value for the key by searching
+// environment variables
+func (c *ConfigParser) GetFromEnv(key string) interface{} {
```
There's no need to export this method as it's only called within this file. Please unexport it.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+	if c.envPrefix != "" {
+		envKey = strings.ToUpper(c.envPrefix + "_" + envKey)
+	}
+	if c.envKeyReplacer != nil {
+		envKey = c.envKeyReplacer.Replace(envKey)
+	}
```
This is where we can inline the call to `strings.ReplacAll`.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -325,9 +553,8 @@ func EnhancedExactUnmarshal(v *viper.Viper, output interface{}) error {
-
-	getterWithClass := func(key string) interface{} { return v.Get(key) } // hide receiver
+	baseKeys := c.GetParsedConfig()
+	getterWithClass := func(key string) interface{} { return c.GetFromEnv(key) }
```
No need for this function. Just pass `c.getFromEnv` (once unexported) directly to `getKeysRecursively`.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,230 @@ import (
+	return os.Getenv(envKey)
+}
+
+// Return the parsed config store
+func (c *ConfigParser) GetParsedConfig() map[string]interface{} {
+	return c.config
+}
```
This function is only used by `EnhancedExactUnmarshal` and can probably be unexported.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -315,7 +543,7 @@ func bccspHook(f reflect.Type, t reflect.Type, data interface{}) (interface{}, e
 // producing error when extraneous variables are introduced and supporting
 // the time.Duration type
-func EnhancedExactUnmarshal(v *viper.Viper, output interface{}) error {
+func EnhancedExactUnmarshal(c *ConfigParser, output interface{}) error {
```
This is no longer a function around viper so it should be converted into a method on `ConfigParser`.

sykesm commented on orderer/common/localconfig/config.go:
```diff
@@ -16,7 +16,7 @@ import (
 	"github.com/hyperledger/fabric/common/viperutil"
 	coreconfig "github.com/hyperledger/fabric/core/config"
-	"github.com/spf13/viper"
+	//"github.com/spf13/viper"
```
please remove commented out code

sykesm reviewed (changes requested):
Given the approach is to stop using viper, we don't need to maintain the generalizations implied by viper.  That means a number of methods can be removed and/or unexported.

Also, since this is not tied to viper, the package should be renamed to something other than `viperutil`.  I'm not sure what the best name so it's probably best to address the other comments first.

Thanks.

nbasker commented:
> Thanks for the PR. Sorry for the delay in reviewing. I've been focusing on other things.
> 
> Looks like the approach you've taken is to extract a bunch of code from viper and inlined it into the `viperutil` package. While normally I'd push back on that because we wouldn't benefit from the enhancements and fixes from that community, but, given we're stuck on an ancient version, it's probably good enough.
> 
> My goal is to do a detailed review over the next day or two. Based on my quick glance, I think there are a few things that were pulled over that we simply don't need and that should be removed (e.g. `SetConfigType`, `AutomaticEnv`), there are some idiomatic Go changes that I'd like to see, and there are tests that need to be backfilled.

Shall be addressing the review comments and raising a PR in the next few days.

nbasker commented:
I still need to take care of a few more changes given, shall post once all the changes are taken care.

nbasker commented:
Addressed all the code review comments and updated the PR.
Regarding the tests that need to be backfilled, should it be around the ConfigParse methods.
Further, instead of viperutil, this can be moved into common/configparse/configparse.go and configparse_test.go to reflect that it is not viperutil anymore (or any other better name).
Since the dependency between viperutil.EnhancedExactUnmarshal() and viper is removed, updating the viper to latest can also be considered.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
+// getParsedConfig returns the parsed config store
+func (c *ConfigParser) getParsedConfig() map[string]interface{} {
+	return c.config
+}
```
This is unexported and only returns a field; we don't need a method to access one internal bit of state with no decorations.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
+		paths = ConfigPaths()
+	}
+	for _, cp := range paths {
+		file := c.searchInPath(cp)
```
We like flat conditions when possible. There's no need for if/else.

```go
paths := c.configPaths
if len(paths) == 0 {
    paths = ConfigPaths()
}

for _, p := range Paths {
...
}
```

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
+
+// Search for the configName in all configPaths
+func (c *ConfigParser) findConfigFile() string {
+
```
nit: we can drop the blank lines after method declarations. There are a few in here.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
 var logger = flogging.MustGetLogger("viperutil")
 
-type viperGetter func(key string) interface{}
+// prototype for get environment variable
+type envGetter func(key string) interface{}
```
This is a detail of getKeys and very far away from it. Please move it close to where it's used.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
 var logger = flogging.MustGetLogger("viperutil")
 
-type viperGetter func(key string) interface{}
+// prototype for get environment variable
```
Common godoc nit. The documentation should be complete sentences when possible. They won't hold up merging but if they're being updated, it's good to ensure they are complete and end with periods.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
+	logger.Debugf("Searching for config in ", in)
+	for _, ext := range supportedExts {
+		fullPath := filepath.Join(in, c.configName+"."+ext)
+		logger.Debugf("Checking for ", fullPath)
```
Same here. I don't see the need but, if we have it, it should be the correct variant.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
+		logger.Debugf("Checking for ", fullPath)
+		_, err := os.Stat(fullPath)
+		if err == nil {
+			logger.Debugf("Found: ", fullPath)
```
Same here. I don't see the need but, if we have it, it should be the correct variant.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
+// Search for the existence of filename for all supported extensions
+func (c *ConfigParser) searchInPath(in string) (filename string) {
+	var supportedExts []string = []string{"yaml", "yml"}
+	logger.Debugf("Searching for config in ", in)
```
I don't believe we need the debug information but, if it's here, it should be the correct variant. Debugf uses a Printf style format string and value and that's not what's here.

Looks like I need to update the linter.

sykesm commented on common/viperutil/config_util.go:
```diff
@@ -24,14 +28,166 @@ import (
+// Search for the configName in all configPaths
+func (c *ConfigParser) findConfigFile() string {
+
+	logger.Debugf("Searching for config in ", c.configPaths)
```
I don't believe this debug is needed.

sykesm reviewed (commented):
It's looking a lot better. Let's get these comments taken care of and we can discuss other changes as follow-on PRs.

sykesm commented:
Looks like I forgot to request that you squash your commits. This change should be contained in a single commit. Thanks.

nbasker commented:
> Looks like I forgot to request that you squash your commits. This change should be contained in a single commit. Thanks.

Tried squashing my commits, but in between had done a merge from master to sync the bugfix branch, in the process messed up the squash. I am going to drop this PR and the branch and raise a fresh one. Shall include all the latest review comments including "file as the reader to ReadConfig" which got missed earlier.
